node/pkg/guardiansigner/amazonkms.go

```diff
@@ -0,0 +1,167 @@
+
+	return &amazonKmsSigner, nil
+}
+
+func (a *AmazonKms) Sign(hash []byte) (signature []byte, err error) {
+
+	// request signing
+	res, err := a.svc.Sign(context.TODO(), &kms.SignInput{
+		KeyId:            aws.String(a.KeyId),
+		Message:          hash,
+		SigningAlgorithm: kms_types.SigningAlgorithmSpecEcdsaSha256,
+		MessageType:      kms_types.MessageTypeDigest,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Signing failed: %w", err)
+	}
+
+	// decode r and s values
+	r, s := derSignatureToRS(res.Signature)
+
+	// if s is greater than secp256k1HalfN, we need to substract secp256k1N from it
+	sBigInt := new(big.Int).SetBytes(s)
+	if sBigInt.Cmp(secp256k1HalfN) > 0 {
+		s = new(big.Int).Sub(secp256k1N, sBigInt).Bytes()
+	}
+
+	// r and s need to be 32 bytes in size
+	r = adjustBufferSize(r)
+	s = adjustBufferSize(s)
+
+	// AWS KMS does not provide the recovery id. But that doesn't matter too much, since we can
+	// attempt recovery id's 0 and 1, and in the process ensure that the signature is valid.
+	expectedPublicKey := a.PublicKey()
+	signature = append(r, s...)
+
+	// try recovery id 0
+	ecSigWithRecid := append(signature, []byte{0}...)
+	pubkey, err := ethcrypto.SigToPub(hash[:], ecSigWithRecid)
+
+	if bytes.Equal(ethcrypto.CompressPubkey(pubkey), ethcrypto.CompressPubkey(&expectedPublicKey)) {
+		return ecSigWithRecid, nil
+	}
+
+	ecSigWithRecid = append(signature, []byte{1}...)
+	pubkey, err = ethcrypto.SigToPub(hash[:], ecSigWithRecid)
+
+	// try recovery id 1
+	if bytes.Equal(ethcrypto.CompressPubkey(pubkey), ethcrypto.CompressPubkey(&expectedPublicKey)) {
+		return ecSigWithRecid, nil
+	}
+
+	return nil, fmt.Errorf("Failed to generate signature")
+}
+
+func (a *AmazonKms) PublicKey() ecdsa.PublicKey {
+	pubKeyOutput, _ := a.svc.GetPublicKey(context.TODO(), &kms.GetPublicKeyInput{
+		KeyId: aws.String(a.KeyId),
+	})
+
+	var asn1Pubkey asn1EcPublicKey
+	_, _ = asn1.Unmarshal(pubKeyOutput.PublicKey, &asn1Pubkey)
+
+	ecdsaPubkey := ecdsa.PublicKey{
+		X: new(big.Int).SetBytes(asn1Pubkey.PublicKey.Bytes[1 : 1+32]),
+		Y: new(big.Int).SetBytes(asn1Pubkey.PublicKey.Bytes[1+32:]),
+	}
+
+	return ecdsaPubkey
+}
+
+func (a *AmazonKms) Verify(sig []byte, hash []byte) (bool, error) {
+	return true, nil
+}
+
+// https://bitcoin.stackexchange.com/questions/92680/what-are-the-der-signature-and-sec-format
+//  1. 0x30 byte: header byte to indicate compound structure
+//  2. one byte to encode the length of the following data
+//  3. 0x02: header byte indicating an integer
+//  4. one byte to encode the length of the following r value
+//  5. the r value as a big-endian integer
+//  6. 0x02: header byte indicating an integer
+//  7. one byte to encode the length of the following s value
+//  8. the s value as a big-endian integer
+func derSignatureToRS(signature []byte) (rBytes []byte, sBytes []byte) {
+	var sigAsn1 asn1EcSig
+	_, err := asn1.Unmarshal(signature, &sigAsn1)
+
+	if err != nil {
+		panic(err)
```

Comment:
Is panicking part of the final design?

Reply:
No, I don't think it should panic here. I removed it.